autode/opt/optimisers/base.py

```diff
@@ -769,8 +771,82 @@ def _best_hessian_updater(self) -> "HessianUpdater":
             "suitable update strategies"
         )
 
+    def plot_optimisation(
+        self,
+        filename: Optional[str] = None,
+        plot_energy: bool = True,
+        plot_rms_grad: bool = True,
+    ) -> None:
+        """
+        Draw the plot of the energies and/or rms_gradients of
+        the optimisation so far
+
+        ----------------------------------------------------------------------
+        Args:
+            filename (str): Name of the file to plot
+            plot_energy (bool): Whether to plot energy
+            plot_rms_grad (bool): Whether to plot RMS of gradient
+        """
+        if self.iteration < 1:
+            logger.warning("Less than 2 points, cannot draw optimisation plot")
+            return None
+
+        if not self.converged:
+            logger.warning(
+                "Optimisation is not converged, drawing a plot "
+                "of optimiser profile until current iteration"
+            )
+
+        filename = (
+            f"{self._species.name}_opt_plot.pdf"
+            if filename is None
+            else str(filename)
+        )
+
+        plot_optimiser_profile(
+            self._history,
+            filename=filename,
+            plot_energy=plot_energy,
+            plot_rms_grad=plot_rms_grad,
+        )
+        return None
+
+    def write_xyz_trajectory(self, filename: Optional[str] = None) -> None:
+        """
+        Write the trajectory of the optimiser in .xyz format
+
+        Args:
+            filename: Name of the trajectory file (xyz)
+        """
+        assert self._species is not None
+        if self.iteration < 1:
+            logger.warning(
+                "Optimiser did no steps, not saving .xyz trajectory"
+            )
+            return None
+
+        filename = (
+            f"{self._species.name}_opt.trj.xyz"
+            if filename is None
+            else str(filename)
+        )
+
+        if os.path.isfile(filename):
+            logger.warning(f"File {filename} exists, overwriting...")
+            os.remove(filename)
+
+        tmp_spc = self._species.new_species()
+        for coord in self._history:
+            tmp_spc.coordinates = coord.to("cart")
+            tmp_spc.energy = coord.e
+            tmp_spc.print_xyz_file(
+                filename=filename,
+                append=True,
+            )
+        return None
+
 
-class _OptimiserHistory(list):
+class OptimiserHistory(UserList):
```

Comment:
@t-young31 UserList maintains the class when you slice or add or do other list operations. So for example, history + history would return another history object, and history[:-2] would also return history object. Whereas previously it would return a list. UserList a wrapper around list basically. The source is here: https://github.com/python/cpython/blob/a9f01448a99c6a2ae34d448806176f2df3a5b323/Lib/collections/__init__.py#L1206.

For polynomial fitting, I needed to obtain `self._history[:-1].minimum` which is why I made that change. It's not needed right now I guess, but it is still better from a programming point of view I think, as you don't get weird errors from wrong type being generated. What do you think?

Reply:
cool – 💯 agree that looks like the right think to use. let's keep this in mind for other list-like classes